src/renderer/base/renderer.cpp

```diff
@@ -121,96 +121,13 @@ IRenderData* Renderer::GetRenderData() const noexcept
         // Last chance check if anything scrolled without an explicit invalidate notification since the last frame.
         _CheckViewportAndScroll();
 
-        if (_currentCursorOptions)
-        {
-            const auto& buffer = _pData->GetTextBuffer();
-            const auto view = buffer.GetSize();
-            const auto coord = _currentCursorOptions->coordCursor;
-
-            // If we had previously drawn a composition at the previous cursor position
-            // we need to invalidate the entire line because who knows what changed.
-            // (It's possible to figure that out, but not worth the effort right now.)
-            if (_compositionCache)
-            {
-                til::rect rect{ 0, coord.y, til::CoordTypeMax, coord.y + 1 };
-                if (view.TrimToViewport(&rect))
-                {
-                    FOREACH_ENGINE(pEngine)
-                    {
-                        LOG_IF_FAILED(pEngine->Invalidate(&rect));
-                    }
-                }
-            }
-            else
-            {
-                const auto lineRendition = buffer.GetLineRendition(coord.y);
-                const auto cursorWidth = _pData->IsCursorDoubleWidth() ? 2 : 1;
-
-                til::rect rect{ coord.x, coord.y, coord.x + cursorWidth, coord.y + 1 };
-                rect = BufferToScreenLine(rect, lineRendition);
+        _invalidateCurrentCursor(); // Invalidate the previous cursor position.
+        _updateCursorInfo();
+        _invalidateCurrentCursor(); // Invalidate the new cursor position.
 
-                if (view.TrimToViewport(&rect))
-                {
-                    FOREACH_ENGINE(pEngine)
-                    {
-                        LOG_IF_FAILED(pEngine->InvalidateCursor(&rect));
-                    }
-                }
-            }
-        }
-
-        _currentCursorOptions = _GetCursorInfo();
+        _invalidateOldComposition();

```

Comment:
shouldn't this come before `_updateCursorInfo`? Since `_updateCursorInfo` can change `_currentCursorOptions.inViewport`